### ISSUE TITLE
Performance Metrics Results

### ISSUE THREAD

**Issue by orouskhani** [MIC-DKFZ/nnDetection#114]
Hello Michael,

Could you please help me for the following questions?

1- Where can I find the result of performance measures such as sensitivity, fp, fn, .... ?  
2- After running the model for the detection task of ADAM dataset, the results_boxes.json is created in val_resutls.
a) So, is this file the result of test data?
b) this file consists of some texts. For example: "FROC _Score_IoU_0.10: 0.75". Could you pls let me know what does it mean?

Thank you very much!
Maysam

**Comment by mibaumgartner** [comment]
Dear @orouskhani,

since nnDetection is not geared towards a specific medical problem (with a single working point -> "so not a single probability threshold") it does not provide cutoff metrics (so metrics which are only computed at a single threshold) in its current form.

All of the scalar metrics are written inside the results_boxes.json file and comprise of the AP and FROC metric. More information about detection metrics, pitfalls and literature resources can be found here: https://arxiv.org/abs/2206.01653
The IoU part of the metric name refers to the IoU threshold used to compute the metric.

Results from the "curve" (i.e. non scalar) metrics can be found in the results_boxes.pkl and the boxes subdirectory. If you are interested in Sens and FP, the FROC curve should be helpful. 

The `val_results` folder contains the results from the validation set while `test_results` would contain results from the test set.

Best,
Michael

**Comment by orouskhani** [comment]
Dear @mibaumgartner 

Thank you very much for the information provided. 

It seems that the content of results_boxes.json is similar to results_boxes.pkl. Isn't it?

Best,
Maysam

**Comment by mibaumgartner** [comment]
The json file only contains the scalar metrics, the pkl file contains the scalar and the curve metrics (e.g. it contains the entire FROC curve and not just the FROC score).

**Comment by orouskhani** [comment]
Thanks for tour reply. 

I have some simple questions too. Just to double check with you. Thx
1- nnDetection uses Retina Unet?
2- What's the loss function? 
3- I run nnDetection for ADAM dataset. The output folder, val_predictions_nii, has some json file for each case. For example, for case 27, the json file contains as follows. There are three numbers, 1-3. What does it mean? Is it the number of aneurysms? Also what is score?
{
    "1": {
        "score": 0.5380976796150208,
        "label": 0,
        "box": [
            34,
            231,
            40,
            240,
            247,
            257
        ]
    },
    "2": {
        "score": 0.6861448884010315,
        "label": 0,
        "box": [
            58,
            265,
            64,
            279,
            279,
            292
        ]
    },
    "3": {
        "score": 0.8369495868682861,
        "label": 0,
        "box": [
            50,
            232,
            57,
            238,
            376,
            384
        ]
    },
 
Regards,
Maysam

**Comment by mibaumgartner** [comment]
Dear @orouskhani,

1) yes nnDetection uses Retina U-Net
2) Bos classification: BCE, Box Regression: GIoU, Segmentation: CE+Soft Dice 

Please refer to our paper for more details about the training and architecture. 

3) each prediction is associated with a corresponding probability score. The json file are generated to add information to the nifty files, the first integer number refers to the index in the nifty file.

**Comment by orouskhani** [comment]
It really helped me. Thx for your response. 

However, in Json file, the box is: "box": [
50,
232,
57,
238,
376,
384
]
},

It indicates the 3d location of two aneurysms? first one is [50, 232,57,] and second is [238, 376, 384]. Is it true?

Regards,
Maysam

**Comment by mibaumgartner** [comment]
Logging is performed via MLFlow (please refer to their documentation for more info).

No, the coordinates are bounding boxes, one box is defined by two points. The format is x1, y1, x2, y2, z1, z2 => the box points would be [50, 232, 376] and [57, 238, 384].

**Comment by orouskhani** [comment]
Thank you very much for your support. Really appreciate your kind help. 

Just as the last question: 
1) in results_boxes.json: "FROC_score_IoU_0.10": "0.75". Could you please what is the meaning of " FROC score"?
2) I see the FROC curve is presented by different IoU. Why we need to calculate FROC in different IoU?

Regards,
Maysam

**Comment by orouskhani** [comment]
Hi Michael,

I hope you are doing well. 

I see the project's output for ADAM dataset as a text file including the center of coordination. 
Is there any result for the diameter of predicted aneurysm?

Best regards,
Maysam

**Comment by mibaumgartner** [comment]
The challenge only expected the center coordinate, so the provided script only computes the center. It should be straight forward to extend the script to return the approximated diameter as well though. Since the prediction is a box and not an ellipse it won't be perfect though and it can only be roughly estimated.

**Comment by tbskbr** [comment]
> Logging is performed via MLFlow (please refer to their documentation for more info).

Hello, the logging metrics for which train or val is not specified correspond to which subset please? Training or validation subset? Eg, FROC_score_IoU_0.10.

Thank you.

**Comment by mibaumgartner** [comment]
The metrics are computed across the validation set -> BUT: during training we only sample patches from the validation, NOT the entire image is predicted since that would take too much time thus the final result in the json file at the end of the training (entire images were predicted for these) are different from the logging results during training

**Comment by tbskbr** [comment]
Thank you !
